### ISSUE TITLE
add golden copies of log output (presence and absence of Warning/Errors) for tests

### ISSUE THREAD

**Issue by ldemailly** [fortio/fortio#295]
To avoid bugs like #294 we should check for log output during e2e and unit tests

In order to avoid those checks to be too brittle we should filter out timestamp, line numbers and level < info and have an easy way to bless the new output as golden when making logging improvements

some of those are features of the log library already

**Comment by rdanielwetan** [comment]
Hi @ldemailly  I'm new to this repo. Can you please guide me to appropriate code part? and a bit of explanation about the issue would be really helpful!

**Comment by ldemailly** [comment]
@rdanielwetan hi ! sorry for the delay, I've been quite busy and thanks for wanting to help

the relevant code here is the _tests

some of the tests - mostly the logger tests, do check the output but most other tests don't

https://github.com/fortio/fortio/blob/master/log/logger_test.go#L84-L87

but the idea there is to generalize so as we're running tests we flag if the warning/errors logged are as expected or not

in a generic way vs 1 by 1

so this may not be a very easy "first issue"

**Comment by sswastik02** [comment]
Hi @ldemailly,
Is this issue still valid ? If yes, I would like to work on it

**Comment by ldemailly** [comment]
yes it is though the logger has moved to fortio/log and it's still not an easy first issue. we should discuss design ideas and requirements before jumping to implementation

**Comment by sswastik02** [comment]
@ldemailly , does it mean this issue should be opened again in `fortio/log` ?

**Comment by ldemailly** [comment]
possibly depending how it would work (or fortio/testscript), but it can start here or in a test repo

**Comment by sswastik02** [comment]
@ldemailly

Could you provide an example (if available) of tests in this or any other Fortio repo that check for specific outputs? I believe we need a generic test utility to verify if warning or error logs are correctly logged. Could you confirm if my understanding is correct?

For example, we could mock a logger using a Go structure (possibly even a file) and have functions that check if specific messages are logged.

**Comment by ldemailly** [comment]
One example is https://github.com/fortio/fortio/blob/a3e1747ec8df4a011c0b2efa3eac9ab80b4ac197/stats/stats_test.go#L48

the rest moved to https://github.com/fortio/log/blob/main/logger_test.go

but it's handcrafted

**Comment by sswastik02** [comment]
Thanks for providing those examples

> but the idea there is to generalize so as we're running tests we flag if the warning/errors logged are as expected or not
> 
> in a generic way vs 1 by 1

Building upon your concept, we could explore the development of a test utility designed to capture output throughout unit tests or end-to-end tests. This utility could actively engage users during test execution, seeking confirmation if the observed output aligns with expected results. If a match is found, the utility could seamlessly integrate it as a test. However, in cases where expectations are not met, the user could conveniently augment the test to align with desired outcomes.

What are your thoughts on this ?

**Comment by ldemailly** [comment]
This could be an inspiration https://github.com/alextanhongpin/testdump
